File: problem_025.go
// -*- compile-command: "go build -o problem_025_go problem_025.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func main() {
        fmt.Println("Project Euler - Problem 25:")
        fmt.Println("What is the index of the first term in the Fibonacci sequence to contain 1000 digits?\n")

        index := 1
	limit := 1000
	digits := 6
        fib_old := 0
        fib_current := 1

        for digits < limit {
                fib_new := fib_current + fib_old
                fib_old = fib_current
                fib_current = fib_new

		// We don't need more than a few digits to be accurate, so
		// keep the number under 6 digits and count how many times
		// we have to shift to know how many decimals we fib has
		for fib_current > 999999 {
			fib_current /= 10
			fib_old /= 10
			digits += 1
		}

		index += 1
        }

        fmt.Println(index)
}

File: problem_025.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_025.swift -o problem_025" -*-
// Copyright (c) 2016 Michael Caldwell

print("Project Euler - Problem 25:")
print("What is the first term in the Fibonacci sequence to contain 1000 digits?\n")

var a: [Int] = [1]
var b: [Int] = [0]
var c: [Int] = [0]
var index = 1

while a.count < 1000+1 {
    c = a

    for i in 0...b.count-1 {
        a[i] += b[i]
    }

    for i in 0...a.count-1 {
        if a.last != 0 { a.append(0) }
        while a[i] > 9{
            a[i+1] += 1
            a[i] -= 10
        }
    }

    b = c
    index += 1
}

print(index)
